common/dev/i2c-mux-pca984x.c

```diff
@@ -0,0 +1,122 @@
+#include <stdio.h>
+#include "sensor.h"
+#include "hal_i2c.h"
+#include "i2c-mux-pca984x.h"
+
+/** 
+ * PCA9846 usage:
+ * PCA9846 only recieves the last byte unless special condition.
+ * The byte data[0] refer to below bit-map; 1 = Enable, 0 = Disable
+ *     Bit 7 6 5 4 3 2 1 0
+ * Channel X X X X 3 2 1 0
+ * 
+ * Example:
+ * File1: plat_i2c.c
+ * K_MUTEX_DEFINE(i2c_bus9_mutex);
+ * 
+ * File2: plat_i2c.h
+ * extern struct k_mutex i2c_bus9_mutex;
+ *
+ * File3: Application
+ * I2C_MSG *msg;
+ * msg->data[0] = 0x1;	// Enable Channel 0 only
+ * msg->data[0] = 0x3;	// Enable Channel 0 + 1
+ * msg->data[0] = 0x4;	// Enable Channel 2 only
+ * msg->data[0] = 0x8;	// Enable Channel 3 only
+ * msg->lock = &i2c_bus9_mutex;
+ * i2c_mux_pca9846_lock(msg);
+ * // Accessing device code here //
+ * i2c_mux_pca9846_unlock(msg);
+ */
+
+/**
+ * @brief lock the specific mutex then enable mux channel.
+ * 
+ * Check lock_count before enable channel.
+ * Set 5 seconds timeout if wait for unlock failed.
+ * After timeout, the mutex will be force re-init to unlock for preventing infinte lock.
+ * Once lock the mutex, switch the channel based on i2c data.
+ * 
+ * @param msg                   i2c message structure.
+ * @param msg->bus              bus number where mux locate.
+ * @param msg->target_addr      slave addr where mux locate.
+ * @param msg->lock             the mutex addr for kernel reference.
+ * @param msg->lock->lock_count number of lock using
+ * 
+ * @retval true  mutex lock and enable mux channel successfully.
+ * @retval false mutex lock or i2c send failed.
+ */
+bool i2c_mux_pca9846_lock(I2C_MSG *msg)
+{
+	if (!msg) {
+		printf("[%s] Received null pointer\n", __func__);
+		return false;
+	}
+
+	uint8_t retry = 3;
+	for (uint8_t i = 1; i <= retry; ++i) {
+		if (msg->lock->lock_count) {
```

Comment:
During review, this was called out as not being thread-safe.  Please rework to make this thread-safe.

Reply:
@garnermic May I know the meaning of thread-safe ? Does that mean force re-initialization (force unlock) is not a safe action ? We implmented this for preventing infinite waiting.

Reply:
By thread-safe, the way to access the mutex isn't to directly access the lock itself.  The expectation is the k_mutex_init is implemented during the initialization and within this function, you would use the k_mutex_lock the corresponding k_mutex_unlock in the unlock function.  For an example of expected usage, please look at the common/dev/pex89000.c source file.

Reply:
@KaidanWu-wiwynn why not just call k_mutex_lock(msg->lock, K_WAIT)?

Reply:
Hi @GoldenBug,
k_mutex_lock does not lock the same thread that different actions still able to access and control it.
For example, IPMB and shell commands both still able to access after did some experiments.

We planned to modify into thread-safe which will remove this portion. Thank you.